firestore/from_value.go

```diff
@@ -32,221 +32,267 @@ func setFromProtoValue(x interface{}, vproto *pb.Value, c *Client) error {
 	return setReflectFromProtoValue(v.Elem(), vproto, c)
 }
 
-// setReflectFromProtoValue sets v from a Firestore Value.
-// v must be a settable value.
-func setReflectFromProtoValue(v reflect.Value, vproto *pb.Value, c *Client) error {
+// setReflectFromProtoValue sets vDest from a Firestore Value.
+// vDest must be a settable value.
+func setReflectFromProtoValue(vDest reflect.Value, vprotoSrc *pb.Value, c *Client) error {
 	typeErr := func() error {
-		return fmt.Errorf("firestore: cannot set type %s to %s", v.Type(), typeString(vproto))
+		return fmt.Errorf("firestore: cannot set type %s to %s", vDest.Type(), typeString(vprotoSrc))
 	}
 
-	val := vproto.ValueType
+	typeErrWithArgs := func(destType string) error {
+		return fmt.Errorf("firestore: cannot set type %s to %s", destType, typeString(vprotoSrc))
+	}
+
+	valTypeSrc := vprotoSrc.ValueType
 	// A Null value sets anything nullable to nil, and has no effect
 	// on anything else.
-	if _, ok := val.(*pb.Value_NullValue); ok {
-		switch v.Kind() {
+	if _, ok := valTypeSrc.(*pb.Value_NullValue); ok {
+		switch vDest.Kind() {
 		case reflect.Interface, reflect.Ptr, reflect.Map, reflect.Slice:
-			v.Set(reflect.Zero(v.Type()))
+			vDest.Set(reflect.Zero(vDest.Type()))
 		}
 		return nil
 	}
 
 	// Handle special types first.
-	switch v.Type() {
+	switch vDest.Type() {
 	case typeOfByteSlice:
-		x, ok := val.(*pb.Value_BytesValue)
+		x, ok := valTypeSrc.(*pb.Value_BytesValue)
 		if !ok {
 			return typeErr()
 		}
-		v.SetBytes(x.BytesValue)
+		vDest.SetBytes(x.BytesValue)
 		return nil
 
 	case typeOfGoTime:
-		x, ok := val.(*pb.Value_TimestampValue)
+		x, ok := valTypeSrc.(*pb.Value_TimestampValue)
 		if !ok {
 			return typeErr()
 		}
 		if err := x.TimestampValue.CheckValid(); err != nil {
 			return err
 		}
-		v.Set(reflect.ValueOf(x.TimestampValue.AsTime()))
+		vDest.Set(reflect.ValueOf(x.TimestampValue.AsTime()))
 		return nil
 
 	case typeOfProtoTimestamp:
-		x, ok := val.(*pb.Value_TimestampValue)
+		x, ok := valTypeSrc.(*pb.Value_TimestampValue)
 		if !ok {
 			return typeErr()
 		}
-		v.Set(reflect.ValueOf(x.TimestampValue))
+		vDest.Set(reflect.ValueOf(x.TimestampValue))
 		return nil
 
 	case typeOfLatLng:
-		x, ok := val.(*pb.Value_GeoPointValue)
+		x, ok := valTypeSrc.(*pb.Value_GeoPointValue)
 		if !ok {
 			return typeErr()
 		}
-		v.Set(reflect.ValueOf(x.GeoPointValue))
+		vDest.Set(reflect.ValueOf(x.GeoPointValue))
 		return nil
 
 	case typeOfDocumentRef:
-		x, ok := val.(*pb.Value_ReferenceValue)
+		x, ok := valTypeSrc.(*pb.Value_ReferenceValue)
 		if !ok {
 			return typeErr()
 		}
 		dr, err := pathToDoc(x.ReferenceValue, c)
 		if err != nil {
 			return err
 		}
-		v.Set(reflect.ValueOf(dr))
+		vDest.Set(reflect.ValueOf(dr))
+		return nil
+
+	case typeOfVector:
```

Comment:
I would break this into its own function and put it near the toProtoValue method so people can see how the two match up. Here is my version, which avoids reflection:
```
func vectorFromProtoValue(v *pb.Value) (Vector, bool) {
        if v == nil {
                return nil, false
        }
        x, ok := v.ValueType.(*pb.Value_MapValue)
        if !ok {
                return nil, false
        }
        m := x.MapValue.Fields
        if vv, ok := stringFromProtoValue(m[typeKey]); !ok || vv != vectorValue {
                return nil, false
        }
        val, ok := m["value"]
        if !ok {
                return nil, false
        }
        av, ok := val.ValueType.(*pb.Value_ArrayValue)
        if !ok {
                return nil, false
        }
        fvals := av.ArrayValue.Values
        f32s := make([]float32, len(fvals))
        for i, fval := range fvals {
                dv, ok := fval.ValueType.(*pb.Value_DoubleValue)
                if !ok {
                        return nil, false
                }
                f32s[i] = float32(dv.DoubleValue)
        }
        return Vector(f32s), true
}

func stringFromProtoValue(v *pb.Value) (string, bool) {
        if v == nil {
                return "", false
        }
        sv, ok := v.ValueType.(*pb.Value_StringValue)
        if !ok {
                return "", false
        }
        return sv.StringValue, true
}
```

Reply:
refactored
returned error instead of boolean to avoid such issues: https://github.com/googleapis/google-cloud-go/issues/10061